### ISSUE TITLE
[sklearn#33793] Optimized _class_means in sklearn/discriminant_analysis.py

### ISSUE THREAD

**Issue by godofecht** [flooooooooooow/flow#378]
**Mirrored from [scikit-learn#33793](https://github.com/scikit-learn/scikit-learn/issues/33793)**

Original title: Optimized _class_means in sklearn/discriminant_analysis.py
Original labels: Performance, module:discriminant_analysis, Needs Investigation
Created: 2026-04-19T14:51:06Z

---

> [!WARNING]
> This issue is not yet ready for a PR. If you are interested in contributing to scikit-learn, please have a look at our [contributing guidelines](https://scikit-learn.org/dev/developers/contributing.html), and in particular the sections for [new contributors](https://scikit-learn.org/dev/developers/contributing.html#new-contributors) and the ["Needs triage"](https://scikit-learn.org/dev/developers/contributing.html#issues-tagged-needs-triage) label.

### Describe the workflow you want to enable

This improves the runtime of solves within the Linear Discriminant Analysis class by speeding up _class_means which is called repeatedly. Please note that this is improving an existing TODO on line 122 of sklearn/discriminant_analysis.py.

### Describe your proposed solution

Based on the existing #TODO that is on line 122 of sklearn/discriminant_analysis.py, I found that the performance of _class_means can be sped up by an average of about 17x by switching from bin count and add.at to one hot matmul. Across 24 test cases it improved all of them by factors varying from 4x to 25x speedups depending on the inputs.

### Describe alternatives you've considered, if relevant

I also tried the combo of merge sort and reduceat for chunk sums. This had only slight improvement and sometimes regressed. My proposed solution showed no regressions in my test suite.

---

## Flow context

This issue is mirrored from scikit-learn to explore whether Flow encounters the same problem or can approach it differently. Add analysis below.

**Comment by godofecht** [comment]
Moved to [20](https://github.com/godofecht/flow-scikit/issues/20) in the flow-scikit repository. Further library-specific triage and implementation will happen there.

**Comment by godofecht** [comment]
Moved to [21](https://github.com/godofecht/flow-scikit/issues/21) in the flow-scikit repository. Further library-specific triage and implementation will happen there.
